<a href="https://colab.research.google.com/github/mikeniemant/QS_shiny/blob/master/tidyverse_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# An introduction to Tidyverse

In this tutorial, we will explore the first four steps of the Tidyverse data science workflow:
1. Import
2. Tidy
3. Transform
4. Visualise
5. Model

Today we will be working with an electronic health records (EHR) dataset. Don't worry about privacy related information, this is synthetic data based on real summary statistics / values from the UMCU.

We will use R in this Google Colab environment. Before we can start, we first have to do three things:
- initiliaze R
- install tidyverse

## 1. Initialize work environment
### 1.1. Initialize R
To run R in this Python environment, we need to load the `rpy2` library. Either click on the play button or put your cursor in the chunk of code below, and press (CMD + ENTER).

In [ ]:
%load_ext rpy2.ipython

The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython


Run the following command to test whether R code is executed.

In [ ]:
%%R
x <- seq(1,5)
x

[1] 1 2 3 4 5


As you can see, Google Colab is instructed to run R code when the 'code chunk' (or 'code cell') starts with the following statement: '%%R'. Otherwise, it will execute Python code.

Feel free to add more chunks of code or text in this notebook, by using one of the following options:
- hover your mouse either above or above a chunk, and click on 'code' or 'text'
- In the menu bar, blick on 'Insert', and then on 'Code cell' / 'Text cell'

### 1.2 Install Tidyverse
Google Colab allows us to install our own packages. Normally it takes quite a while before the complete `Tidyverse` meta package is installed, apparently Colab has it already pre-installed somewhere. Lucky us!

In [ ]:
%%R
install.packages("tidyverse")

R[write to console]: Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

R[write to console]: trying URL 'https://cran.rstudio.com/src/contrib/tidyverse_1.3.1.tar.gz'

R[write to console]: Content type 'application/x-gzip'
R[write to console]:  length 702779 bytes (686 KB)

R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[write to console]: =
R[wr

Import the `tidyverse` package and test it by running the following chunk.

In [ ]:
%%R
library(tidyverse)
mtcars %>% 
  nrow()

[1] 32


`32` as output? Perfect!

## 2. Tidyverse data science workflow
### 2.1 Import dataset
Now we will start importing the dataset from a Github repository I created: https://github.com/mikeniemant/ehr_tutorial.

First, have a look at the two files in the repo:
- ehr.csv: this is the dataset
- data_dic.csv: this is dictionary of all variables

We can directly download the EHR data from the Github repo with the `read_csv` function from the `readr` package.

In [ ]:
%%R
dat <- read_csv("https://github.com/mikeniemant/ehr_tutorial/raw/main/ehr.csv")

Rows: 384 Columns: 18
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr   (1): sx
dbl  (15): study_id, outcome, age, hr, sbp, rr, scr, white_blood_cell_count,...
date  (1): visit_date
time  (1): visit_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


The `read_csv` automatically assigns a class to each column. Let's have a look at the dataset.

In [ ]:
%%R
dat

# A tibble: 384 × 18
   study_id outcome   age sx     visit_date visit_time    hr   sbp    rr   scr
      <dbl>   <dbl> <dbl> <chr>  <date>     <time>     <dbl> <dbl> <dbl> <dbl>
 1      244       1  78.3 male   2020-01-31 13:12:31    84.4  96.1  17.6  87.1
 2      199       1  86.8 female 2020-03-05 16:07:02    56.4  97.7  15.6  86.0
 3      271       1  68.2 male   2020-05-31 00:00:00    95.3  92.5  21.5  74.9
 4      101       0  77.6 male   2020-02-14 10:29:08    88.4 135.   15.3  91.3
 5      261       1  72.7 male   2020-02-15 00:34:27    58.1 108.   12.3  80.8
 6      321       1  52.4 female 2020-05-23 16:02:02    65.7  95.3  14.4  98.5
 7      171       0  56.2 male   2020-04-30 01:16:08    88.3  76.3  24.7  96.7
 8       70       0  70.4 female 2020-02-17 14:39:47    73.5 114.   24.2  95.2
 9       40       1  60.2 female 2020-04-14 18:51:32    71.0 118.   23.2  84.8
10       45       0  39.3 male   2020-03-21 13:00:18    80.6  89.6  16.0  85.7
# … with 374 more rows, and 8 m

As we have imported the data into an tibble, R only returns the first ten rows (instead of all 384) and specifies the class of each column beneath the column name (chr, dbl, etc.). 

To get a readable overview, and depending on your screen width, some columns may be removed from the  output.

This dataset consists out of four data types:
- demographic
  - age (numeric)
  - sex (categorical), 0 = female, 1 = male
- hospital visit
  - date of visit (date)
  - time of visit (time)
  - department (categorical, X values)
- vital
  - heart rate (hr), (numeric)
  - systolic blood pressure (sbp) (numeric)
  - diastolic blood pressure
  - respiratory rate (rr) (numeric)
- laboratory
  - serum creatinine (scr)
  - white blood cell count (wbc)

Take a look at the dataset for a few second, do you notice some errors? Is this a tidy dataset?

It looks like there is a typo. The fourth column should be 'sex' instead of 'sx'. We can use the `rename` function from the `dplyr` package

In [ ]:
%%R
dat <- dat %>%
  rename(sex = sx)

Did it work? Run the next chunk

In [ ]:
%%R
dat

# A tibble: 384 × 18
   study_id outcome   age sex    visit_date visit_time    hr   sbp    rr   scr
      <dbl>   <dbl> <dbl> <chr>  <date>     <time>     <dbl> <dbl> <dbl> <dbl>
 1      244       1  78.3 male   2020-01-31 13:12:31    84.4  96.1  17.6  87.1
 2      199       1  86.8 female 2020-03-05 16:07:02    56.4  97.7  15.6  86.0
 3      271       1  68.2 male   2020-05-31 00:00:00    95.3  92.5  21.5  74.9
 4      101       0  77.6 male   2020-02-14 10:29:08    88.4 135.   15.3  91.3
 5      261       1  72.7 male   2020-02-15 00:34:27    58.1 108.   12.3  80.8
 6      321       1  52.4 female 2020-05-23 16:02:02    65.7  95.3  14.4  98.5
 7      171       0  56.2 male   2020-04-30 01:16:08    88.3  76.3  24.7  96.7
 8       70       0  70.4 female 2020-02-17 14:39:47    73.5 114.   24.2  95.2
 9       40       1  60.2 female 2020-04-14 18:51:32    71.0 118.   23.2  84.8
10       45       0  39.3 male   2020-03-21 13:00:18    80.6  89.6  16.0  85.7
# … with 374 more rows, and 8 m

Bravo! 
I always find long column names annoying.. can you change 'white_blood_cell_count' to 'wbc'? Double click on the chunk below and write your code. Make sure to not dletee the '%%R', otherwise your code will not run! 

In [ ]:
%%R
YOUR R CODE HERE

In [ ]:
#@title
%%R
dat <- dat %>%
  rename(wbc = white_blood_cell_count)

Grazie mille! The study_ids are not in order.. Patients can visits the hospital multiple times within one year. We can sort on multiple columns with the `arrange` command.

In [ ]:
%%R
dat <- dat %>%
  arrange(study_id, visit_date, visit_time)

In [ ]:
%%R
dat

# A tibble: 384 × 18
   study_id outcome   age sex    visit_date visit_time    hr   sbp    rr   scr
      <dbl>   <dbl> <dbl> <chr>  <date>     <time>     <dbl> <dbl> <dbl> <dbl>
 1        1       0  80.1 male   2020-03-23 03:05:38    76.1 115.   20.5  88.8
 2        2       1  51.2 male   2020-02-23 03:58:26    75.9 116.   18.8  92.0
 3        3       1  65.1 male   2020-03-13 05:02:20    74.7 150    20.6  72.8
 4        3       0  65.1 male   2020-04-21 22:15:30    74.7 150    20.6  72.8
 5        4       1  69.1 male   2020-02-16 19:37:07    86.3 108.   18.0  72.3
 6        5       1  65.7 male   2020-04-02 11:40:15    96.5 137.   14.4  78.4
 7        6       1  58.1 male   2020-03-14 09:36:20    87.5 109.   19.6  93.8
 8        9       1  89.8 female 2020-01-04 23:13:18    85.0 116.   18.2  84.9
 9        9       0  89.8 male   2020-05-23 06:30:42    85.0 116.   18.2  84.9
10       10       0  58.7 male   2020-01-05 08:44:38    71.6  87.5  17.9  90.6
# … with 374 more rows, and 8 m

FIX THE STUDY IDs

### 2.2 Tidy

Let's have a look at the department columns. We can use the `select` command to make a selection without assigning it to the `dat` object. There are multiple options to get the same result, run the following chunks

In [ ]:
%%R
dat %>% 
  select(study_id, oncology, obstetrics, neurology, nephrology, 
         internal_medicine, cardiology, hematology)

# A tibble: 384 × 8
   study_id oncology obstetrics neurology nephrology internal_medici… cardiology
      <dbl>    <dbl>      <dbl>     <dbl>      <dbl>            <dbl>      <dbl>
 1      244        1         NA        NA         NA               NA         NA
 2      199       NA          1        NA         NA               NA         NA
 3      271       NA          1        NA         NA               NA         NA
 4      101       NA         NA         1         NA               NA         NA
 5      261       NA         NA        NA          1               NA         NA
 6      321       NA          1        NA         NA               NA         NA
 7      171       NA         NA        NA         NA                1         NA
 8       70       NA         NA        NA          1               NA         NA
 9       40       NA         NA         1         NA               NA         NA
10       45       NA         NA        NA         NA               NA          1
# … with

In [ ]:
%%R
dat %>% 
  select(study_id, 12:18)

# A tibble: 384 × 8
   study_id oncology obstetrics neurology nephrology internal_medici… cardiology
      <dbl>    <dbl>      <dbl>     <dbl>      <dbl>            <dbl>      <dbl>
 1      244        1         NA        NA         NA               NA         NA
 2      199       NA          1        NA         NA               NA         NA
 3      271       NA          1        NA         NA               NA         NA
 4      101       NA         NA         1         NA               NA         NA
 5      261       NA         NA        NA          1               NA         NA
 6      321       NA          1        NA         NA               NA         NA
 7      171       NA         NA        NA         NA                1         NA
 8       70       NA         NA        NA          1               NA         NA
 9       40       NA         NA         1         NA               NA         NA
10       45       NA         NA        NA         NA               NA          1
# … with

Phoe.. This does not look tidy. We can fix this but pivoting the data from a wide to a long format. Before we make any errors, create a new object called `departments` by selecting the `study_id` and all `department` columns.

In [ ]:
%%R
YOUR R CODE HERE

In [ ]:
#@title
%%R
departments <- dat %>%
  select(1, 12:18)

In [ ]:
%%R
departments %>% 
  pivot_longer(cols = -1, names_to = "deparment", values_to = "value")

# A tibble: 2,688 × 3
   study_id deparment         value
      <dbl> <chr>             <dbl>
 1      244 oncology              1
 2      244 obstetrics           NA
 3      244 neurology            NA
 4      244 nephrology           NA
 5      244 internal_medicine    NA
 6      244 cardiology           NA
 7      244 hematology           NA
 8      199 oncology             NA
 9      199 obstetrics            1
10      199 neurology            NA
# … with 2,678 more rows


In [ ]:
%%R
departments %>% 
  pivot_longer(cols = -1, names_to = "deparment", values_to = "value") %>%
  na.omit()

# A tibble: 384 × 3
   study_id deparment         value
      <dbl> <chr>             <dbl>
 1      244 oncology              1
 2      199 obstetrics            1
 3      271 obstetrics            1
 4      101 neurology             1
 5      261 nephrology            1
 6      321 obstetrics            1
 7      171 internal_medicine     1
 8       70 nephrology            1
 9       40 neurology             1
10       45 cardiology            1
# … with 374 more rows


In [ ]:
%%R
dat <- dat %>%
  pivot_longer(cols = 12:18, names_to = "deparment", values_to = "value") %>%
  na.omit()

In [ ]:
%%R
dat

# A tibble: 384 × 13
   study_id outcome   age sex    visit_date visit_time    hr   sbp    rr   scr
      <dbl>   <dbl> <dbl> <chr>  <date>     <time>     <dbl> <dbl> <dbl> <dbl>
 1        1       0  80.1 male   2020-03-23 03:05:38    76.1 115.   20.5  88.8
 2        2       1  51.2 male   2020-02-23 03:58:26    75.9 116.   18.8  92.0
 3        3       1  65.1 male   2020-03-13 05:02:20    74.7 150    20.6  72.8
 4        3       0  65.1 male   2020-04-21 22:15:30    74.7 150    20.6  72.8
 5        4       1  69.1 male   2020-02-16 19:37:07    86.3 108.   18.0  72.3
 6        5       1  65.7 male   2020-04-02 11:40:15    96.5 137.   14.4  78.4
 7        6       1  58.1 male   2020-03-14 09:36:20    87.5 109.   19.6  93.8
 8        9       1  89.8 female 2020-01-04 23:13:18    85.0 116.   18.2  84.9
 9        9       0  89.8 male   2020-05-23 06:30:42    85.0 116.   18.2  84.9
10       10       0  58.7 male   2020-01-05 08:44:38    71.6  87.5  17.9  90.6
# … with 374 more rows, and 3 m

In [ ]:
%%R
dat <- dat %>%
  select(-value)

In [ ]:
%%R
dat

# A tibble: 384 × 12
   study_id outcome   age sex    visit_date visit_time    hr   sbp    rr   scr
      <dbl>   <dbl> <dbl> <chr>  <date>     <time>     <dbl> <dbl> <dbl> <dbl>
 1        1       0  80.1 male   2020-03-23 03:05:38    76.1 115.   20.5  88.8
 2        2       1  51.2 male   2020-02-23 03:58:26    75.9 116.   18.8  92.0
 3        3       1  65.1 male   2020-03-13 05:02:20    74.7 150    20.6  72.8
 4        3       0  65.1 male   2020-04-21 22:15:30    74.7 150    20.6  72.8
 5        4       1  69.1 male   2020-02-16 19:37:07    86.3 108.   18.0  72.3
 6        5       1  65.7 male   2020-04-02 11:40:15    96.5 137.   14.4  78.4
 7        6       1  58.1 male   2020-03-14 09:36:20    87.5 109.   19.6  93.8
 8        9       1  89.8 female 2020-01-04 23:13:18    85.0 116.   18.2  84.9
 9        9       0  89.8 male   2020-05-23 06:30:42    85.0 116.   18.2  84.9
10       10       0  58.7 male   2020-01-05 08:44:38    71.6  87.5  17.9  90.6
# … with 374 more rows, and 2 m

### 2.3 Transform
Looking at `study_id` 9, the age for this person is the same for both visits. We can round all values in this column with the `mutate` command.

In [ ]:
%%R
dat <- dat %>%
  mutate(age = round(age))

In [ ]:
%%R
dat

# A tibble: 384 × 12
   study_id outcome   age sex    visit_date visit_time    hr   sbp    rr   scr
      <dbl>   <dbl> <dbl> <chr>  <date>     <time>     <dbl> <dbl> <dbl> <dbl>
 1        1       0    80 male   2020-03-23 03:05:38    76.1 115.   20.5  88.8
 2        2       1    51 male   2020-02-23 03:58:26    75.9 116.   18.8  92.0
 3        3       1    65 male   2020-03-13 05:02:20    74.7 150    20.6  72.8
 4        3       0    65 male   2020-04-21 22:15:30    74.7 150    20.6  72.8
 5        4       1    69 male   2020-02-16 19:37:07    86.3 108.   18.0  72.3
 6        5       1    66 male   2020-04-02 11:40:15    96.5 137.   14.4  78.4
 7        6       1    58 male   2020-03-14 09:36:20    87.5 109.   19.6  93.8
 8        9       1    90 female 2020-01-04 23:13:18    85.0 116.   18.2  84.9
 9        9       0    90 male   2020-05-23 06:30:42    85.0 116.   18.2  84.9
10       10       0    59 male   2020-01-05 08:44:38    71.6  87.5  17.9  90.6
# … with 374 more rows, and 2 m

Create two additional columns:
- ratio sbp - dbp (`sbp_dbp_ratio`): sbp / dbp
- mean arterial blood pressure (`map`): (2 x diastolic blood pressure + systolic blood pressure) / 3


In [ ]:
%%R
YOUR CODE HERE

In [ ]:
#@title
%%R
dat <- dat %>%
  mutate(sbp_dbp_ratio = sbp / dbp
         map = (2 x dbp + sbp) / 3)

## 2.4 Visualise
